Add -SecurityDescriptorSddl parameter to Set-Service

kvprasoon opened the pull request:
## PR Summary

<!-- summarize your PR between here and the checklist -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link:https://github.com/PowerShell/PowerShell/issues/5690
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

kvprasoon committed:
Add -SecurityDescriptor parameter to Set-Service

kvprasoon committed:
Define psecurityDescriptor

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
change in status variable

kvprasoon commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1870,6 +1889,37 @@ protected override void ProcessRecord()
+                                                    securityDescriptor,
+                                                    NativeMethods.stringSDRevision,
+                                                    psecurityDescriptor,
+                                                    IntPtr.Zero);
```
I have no experience in c# and whatever done so far is by referring the docs available. But now I'm stuck and need help here to use the value from `psecurityDescriptor` to pass to `SetServiceObjectSecurity` method as a `DWORD`.

SteveL-MSFT replied:
In this case, perhaps you should use https://docs.microsoft.com/en-us/dotnet/api/system.security.accesscontrol.rawsecuritydescriptor.-ctor?redirectedfrom=MSDN&view=netcore-2.1#System_Security_AccessControl_RawSecurityDescriptor__ctor_System_String_

iSazonov replied:
Perhaps this helps https://stackoverflow.com/questions/15771998/how-to-give-a-user-permission-to-start-and-stop-a-particular-service-using-c-sha

kvprasoon replied:
Thanks @iSazonov , that gave me more clarity. Will setting only DACL suffice here ? how about SACL , Owner, Group ?

iSazonov replied:
I reviewed the example from my link above and found it being very informative. There is "psd" (security descriptor) which contains all you ask.

kvprasoon replied:
Thanks, I am able to set the DACL. Currently finding way to use `SE_SECURITY_NAME` privilege for getting `ACCESS_SYSTEM_SECURITY` permission in the handle.

SteveL-MSFT replied:
DACL is the actual permission, SACL is for auditing.  You would want to support both if both are given.  But it would also be ok to just document that only DACL is supported for now.

Look online for enabling security privilege in C#.

kvprasoon replied:
Thanks @SteveL-MSFT . I looked into few docs online and found `AdjustTokenPrivileges` , but having some difficulty in using it w.r.t few parameters. This is becoming a great learning for me !

kvprasoon replied:
I have added change for only DACL now a suggested. Will open an issue for SACL support to track. I will update the Docs once review is done.

iSazonov replied:
```
sc.exe sdshow wuauserv

D:(A;;CCLCSWRPLORC;;;AU)(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;BA)(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;SY)
```

kvprasoon replied:
My question was on which service should I take to change the setting ? I tested it by creating a dummy service in my box.

iSazonov replied:
You could do the same here (feature tagged test).

kvprasoon replied:
Sure. Seems like because of https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1#L19 , tests can fail. I could raise a PR fix this first .

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1560,28 @@ public ServiceStartupType StartupType
+                securityDescriptor = value;
+            }
+        }
+        private string securityDescriptor ;
```
Please use auto-property:
```c#
public string SecurityDescriptor
{
    get;
    set;
}

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2567,7 +2617,10 @@ internal static class NativeMethods
+        internal const DWORD WRITE_DAC =18;
+        internal const DWORD WRITE_OWNER =19;
+        internal const DWORD ACCESS_SYSTEM_SECURITY =24;
+        internal const DWORD stringSDRevision = 1;
```
Why another style?

kvprasoon replied:
I will change it to `SDDL_REVISION_1`

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
DACL Support only

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1561,24 @@ public ServiceStartupType StartupType
 
+        /// <summary>
+        /// The following is the definition of the input parameter "SecurityDescriptor".
+        /// Changes the SecurityDescriptor of the service.
```
From the comment it is not clear that is the security descriptor and its format.

kvprasoon replied:
Will this suffice ?
```
        /// <summary>
        /// The following is the definition of the input parameter "SecurityDescriptor".
        /// Changes the SecurityDescriptor of the service and is accepted in  SDDL format.
        /// </summary>
```

iSazonov replied:
Need approve from English native.

SteveL-MSFT replied:
The first sentence is not necessary.  The second sentence should be "Set the SecurityDescriptor..."

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1561,24 @@ public ServiceStartupType StartupType
+        /// </summary>
+        [Parameter]
+        [Alias("sd")]
+        [ValidateNotNullOrEmpty]
```
We _could_ think about special validator for the security descriptor to help users in interactive sessions.

kvprasoon replied:
using `ValidatePattern`

iSazonov replied:
My comment is only for discussion and not block the PR.
SDDL is not simple. I think we need help users. If it is not validate attribute then it could be a suggestion tip.

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1870,6 +1886,37 @@ protected override void ProcessRecord()
+                            exception,
+                            "CouldNotSetServiceSecurityDescriptor",
+                            ServiceResources.CouldNotSetServiceSecurityDescriptor,
+                            ErrorCategory.PermissionDenied);
```
I am not sure about PermissionDenied - returned status code can be ERROR_INVALID_PARAMETER and others.

kvprasoon replied:
How about InvalidOperation ?

kvprasoon replied:
How about InvalidArgument ?

iSazonov replied:
I am thinking about some catch blocks not it is not request to change - only for discussion.

iSazonov commented:
There is an Issue for gMSA. @kvprasoon Did you investigations for this?

PaulHigin reviewed (changes requested):
This is very useful, but I feel it is not complete and should be added to New-Service as well.  Also the user should be able to verify the sddl on the service with Get-Service.  Is it possible to obtain the service sddl and display it as a property on the returned ServiceController object?

kvprasoon commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1561,24 @@ public ServiceStartupType StartupType
 
+        /// <summary>
+        /// The following is the definition of the input parameter "SecurityDescriptor".
+        /// Changes the SecurityDescriptor of the service.
```
Perfect. Initially I had thoughts on doing this in `Get/New-Service` cmdlets as well, but I never did c# so thought of doing it in `Set-Service` first then to have a separate PR for other two cmdlets.

kvprasoon commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1561,24 @@ public ServiceStartupType StartupType
+        [Parameter]
+        [Alias("sd")]
+        [ValidateNotNullOrEmpty]
+        public string SecurityDescriptor
```
Thanks, I forgot to check for similar existing ones .

kvprasoon commented:
> There is an Issue for gMSA. @kvprasoon Did you investigations for this?

I saw it and will go through it.

kvprasoon commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1870,6 +1886,37 @@ protected override void ProcessRecord()
+                    byte[] securityDescriptorByte = new byte[rawSecurityDescriptor.BinaryLength];
+                    rawSecurityDescriptor.GetBinaryForm(securityDescriptorByte, 0);
+
+                    if(!string.IsNullOrEmpty(SecurityDescriptor))
```
Yep, but this is not a mandatory parameter, absence of this will throw exception from `SetServiceObjectSecurity()`

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
Addressing review comments

SteveL-MSFT commented:
@PaulHigin can you re-review?  Looks like @kvprasoon addressed your concerns

kvprasoon committed:
Merge branch 'master' into srv-sdset

SteveL-MSFT commented:
@kvprasoon for validation, you can use the CIM cmdlets to call the WMI instance method Win32_Service.GetSecurityDescriptor() to verify the SD was changed correctly.  I believe there's an existing test that creates a test service so you can reuse that code.

kvprasoon committed:
add unit tests

PaulHigin commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1870,6 +1886,34 @@ protected override void ProcessRecord()
+                        byte[] rawDacl = new byte[discretionaryAcl.BinaryLength];
+                        discretionaryAcl.GetBinaryForm(rawDacl, 0);
+                        rawSecurityDescriptor.DiscretionaryAcl = new RawAcl(rawDacl, 0);
+                        byte[] securityDescriptorByte = new byte[rawSecurityDescriptor.BinaryLength];
```
For safety, I think we should pass a pointer from a pinned handle for any data we pass to native APIs.  
```
hDacl = GCHandle.Alloc(rawDacl, GCHandleType.Pinned);
lpDacl = hDaclAddrOfPinnedObject();
...
hDacl.Value.Free();
```

kvprasoon replied:
Never used pointers, will understand about the pinning handle and update. Thanks.

kvprasoon replied:
When I use the above code and pass  the lpDacl to
`SetServiceObjectSecurity(
hService,
SecurityInfos.DiscretionaryAcl,
lpDacl)` method, the build passes, but the console crashes when the cmdlet is executed for this use case.

IISResetMe replied:
@PaulHigin what's the purpose of pinning the handle and then immediately passing and freeing it?

PaulHigin replied:
Just to ensure the GC doesn't move data while native apps access it.

kvprasoon replied:
@PaulHigin , with the help of @IISResetMe I have made the changes as per the comments, requesting your review.

IISResetMe replied:
@PaulHigin if it's to enforce common usage throughout the code base, then fair enough, but it is wholly unnecessary in this case - (non-nested one-dimensional arrays of blittable types are implicitly pinned for the duration of the marshalled transaction)[https://github.com/dotnet/corefx/blob/103ef55a43b93815277a0b6771e19675730e8ab7/Documentation/coding-guidelines/interop-pinvokes.md#blittable-types], so passing `byte[]` without explicitly wrapping it in a pinned handle would be just as safe

iSazonov replied:
I think @PaulHigin request _explicit_ pin that is more safe.

PaulHigin replied:
@IISResetMe  I wasn't aware that blittable types are now implicitly pinned ... I was only advocating the pinning of blittable types:

`An instance with nonprimitive (non-blittable) members cannot be pinned. `

@kvprasoon , it looks like this explicit handle pinning is no longer necessary.  Please remove it.  Sorry for the churn!

PaulHigin replied:
@IISResetMe  Thanks for the correction!

IISResetMe replied:
You're most welcome @PaulHigin! 

It did get me thinking though, does it make sense to provide a common wrapper for the pinned handle pattern for use in all pinvoke cases where we pass non-blittable types though? Something like this https://gist.github.com/IISResetMe/1b428db41658cef50f51bc168aee3cc1

PaulHigin replied:
I like the wrapper ... very cool and useful.

PaulHigin reviewed (changes requested):
Please add tests to verify that the security descriptor is successfully added.  You can use WMI:
```
Win32_Serivce.GetSecurityDescriptor()
```

Also, as I mentioned before, I feel this is not a complete user experience.  We should give the user a way to verify that the security descriptor is what they intended.  A user can use WMI to verify service SD, but I think it would be better if the SD was part of the service object (as a property) we returned from Get-Service.  Or perhaps another cmdlet.  But this can be addressed in a separate Issue.  Can you please create a new Issue for this?

It looks like we can use the native API
```
QueryServiceObjectSecurity()
```

kvprasoon commented:
@PaulHigin Thank, I have created #8872 for `Get-Service` . I will have the changes done for the review comments in upcoming weekends.

kvprasoon commented:
> @kvprasoon for validation, you can use the CIM cmdlets to call the WMI instance method Win32_Service.GetSecurityDescriptor() to verify the SD was changed correctly.  I believe there's an existing test that creates a test service so you can reuse that code.
`GetSecurityDescriptor()` method returns a returnvalue, hence I though using `SC SDSHOW`  as it will give the exact output which can be compared with the input.

kvprasoon committed:
using pinned handle

kvprasoon committed:
Merge branch 'master' into srv-sdset

IISResetMe committed:
Fix sddl handle in Set-Service

Alloc pinned handle for `securityDescriptorByte` rather than `rawDacl`, change error handling behavior to write a non-terminating error on failure (throwing exceptions is no good here, will be swallowed by outer tryf block) and finally attempt to safely free the pinned handle

kvprasoon committed:
Merge pull request #2 from IISResetMe/srv-sdset

Fix sddl handle in Set-Service

PaulHigin commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1901,22 +1899,33 @@ protected override void ProcessRecord()
+                            {
+                                int lastError = Marshal.GetLastWin32Error();
+                                Win32Exception exception = new Win32Exception(lastError);
+                                bool accessDenied = exception.NativeErrorCode == 0x5;
```
Please make `0x5` a constant with a descriptive error name.

PaulHigin commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1901,22 +1899,33 @@ protected override void ProcessRecord()
+                                    exception,
+                                    nameof(ServiceResources.CouldNotSetServiceSecurityDescriptorSddl),
+                                    StringUtil.Format(ServiceResources.CouldNotSetServiceSecurityDescriptorSddl, Name, exception.Message),
+                                    accessDenied ? ErrorCategory.PermissionDenied : ErrorCategory.WriteError);
```
I think default error category should be `ErrorCategory.InvalidOperation` rather than `ErrorCategory.WriteError`.

PaulHigin reviewed (changes requested):
Overall it looks good!  Just a couple of minor comments.

PaulHigin reviewed (changes requested):
It looks like we should remove the handle pinning for dotnet Core.  I'll also create an Issue to remove it from RemoteSessionNamedPipe code.

kvprasoon committed:
add unit tests

kvprasoon commented:
@PaulHigin I have updated with required changes and unit tests and CI has failed, seems like not related to this commit.

kvprasoon committed:
review comment change for error msg and const code

kvprasoon committed:
Merge branch 'master' into srv-sdset

PaulHigin commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -9,14 +9,16 @@ Describe "Set/New/Remove-Service cmdlet tests" -Tags "Feature", "RequireAdminOnW
             $password = ConvertTo-SecureString $testPass -AsPlainText -Force
             $creds = [pscredential]::new(".\$userName", $password)
 
+            $svcbinaryname = New-Item -Path TestDrive:\TestExecutable.exe -ItemType File
```
I don't understand why this was changed.

kvprasoon replied:
Because `Get-Command $svcbinaryname` fails as there is no `TestService` executable available. Will this be present in build agent ? Added this change when doing testing locally ?

PaulHigin replied:
Yes, as I understand it, the TestService is part of the test set-up and you should revert this change.

@adityapatwardhan  can you confirm?

PaulHigin commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -72,12 +74,32 @@ Describe "Set/New/Remove-Service cmdlet tests" -Tags "Feature", "RequireAdminOnW
+        Set-Service -Name $TestServiceName1 -SecurityDescriptor $SecurityDescriptorSddl
+        $Counter      = 0
+        $ExpectedSDDL = ConvertFrom-SddlString -Sddl $SecurityDescriptorSddl
+        $UpdatedSDDL  = ConvertFrom-SddlString -Sddl (sc sdshow $TestServiceName1)[-1]
```
Why are you using [-1] instead of [1] here?  Please add comment to clarify.

kvprasoon replied:
Was just  a preference. [1] will work and updated the same with comments.

PaulHigin reviewed (changes requested):
Sorry for the delay.  The changes look good except for some of the test changes.

PoshChan commented:
@TravisEz13, all requests start with the magic word `Please`.

PoshChan commented:
@TravisEz13, did not find any test failures or test failures were not published to results

PoshChan commented:
@TravisEz13, did not find any test failures or test failures were not published to results

PoshChan commented:
@TravisEz13, did not find any test failures or test failures were not published to results

PoshChan commented:
@TravisEz13, your last commit had 3 failures in `PowerShell-CI-windows`
### Set/New/Remove-Service cmdlet tests.Set-Service test cases on the services with dependent relationship.Set-Service can stop a service with dependency
  Error: Expected exactly 'Running', but got Stopped.
```stacktrace
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 374
374:             (Get-Service $testservicename1).Status | Should -BeExactly "Running"
```
### Set/New/Remove-Service cmdlet tests.Set-Service test cases on the services with dependent relationship.Set-Service cannot stop a service with running dependent service
  Error: Expected exactly 'Running', but got Stopped.
```stacktrace
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 374
374:             (Get-Service $testservicename1).Status | Should -BeExactly "Running"
```
### Set/New/Remove-Service cmdlet tests.Set-Service test cases on the services with dependent relationship.Set-Service can stop a service with running dependent service by parameter -Force
  Error: Expected exactly 'Running', but got Stopped.
```stacktrace
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 374
374:             (Get-Service $testservicename1).Status | Should -BeExactly "Running"
```

kvprasoon committed:
adding comment as per review

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
adding comment as per review

kvprasoon committed:
adding comment as per review

kvprasoon committed:
revert dummy service executable creation

kvprasoon commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1561,24 @@ public ServiceStartupType StartupType
+        [Parameter]
+        [Alias("sd")]
+        [ValidateNotNullOrEmpty]
+        public string SecurityDescriptorSddl
```
The same is used for `Enable/Set/Register-PSSessionConfiguration` cmdlets. hence used this name as per @PaulHigin 's suggestion.

https://github.com/PowerShell/PowerShell/pull/8626#discussion_r250320136

SteveL-MSFT replied:
I agree it's redundant, but since there's precedent already it's better to be consistent

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1560,10 +1561,24 @@ public ServiceStartupType StartupType
 
+        /// <summary>
+        /// The following is the definition of the input parameter "SecurityDescriptorSddl".
+        /// Sets the SecurityDescriptorSddl of the service using a SDDL string.
```
The public docs must start with "Gets or sets ...".
I suggest to remove line 1570 at all.

kvprasoon replied:
Removed

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -1870,6 +1886,38 @@ protected override void ProcessRecord()
                         }
                     }
 
+                    // Handle the '-SecurityDescriptorSddl' parameter
```
Please remove obvious comment.

kvprasoon replied:
Removed

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2688,6 +2738,15 @@ internal static extern
+
+        [DllImport(PinvokeDllNames.SetServiceObjectSecurityDllName, CharSet = CharSet.Unicode, SetLastError = true)]
+        internal static extern
+        bool SetServiceObjectSecurity(
```
Please add 
`[return: MarshalAs(UnmanagedType.Bool)]`
after line 2742.
(It is default but explicit is better)

kvprasoon replied:
Done

iSazonov commented on src/System.Management.Automation/utils/PInvokeDllNames.cs:
```diff
@@ -137,5 +137,6 @@ internal static class PinvokeDllNames
         internal const string DeleteServiceDllName = "api-ms-win-service-management-l1-1-0.dll";             /*124*/
         internal const string QueryServiceConfigDllName = "api-ms-win-service-management-l2-1-0.dll";        /*125*/
         internal const string QueryServiceConfig2DllName = "api-ms-win-service-management-l2-1-0.dll";       /*126*/
+        internal const string SetServiceObjectSecurityDllName = "api-ms-win-service-management-l2-1-0.dll";  /*127*/
```
Is it present on all supported platforms?
/cc @TravisEz13 @PaulHigin

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -9,13 +9,15 @@ Describe "Set/New/Remove-Service cmdlet tests" -Tags "Feature", "RequireAdminOnW
         if ($IsWindows) {
             $userName = "testuserservices"
             $testPass = "Secret123!"
```
@TravisEz13 Can we use a clear text password?

SteveL-MSFT replied:
Even in tests, we should generate a password.

kvprasoon replied:
@TravisEz13  will this function be available in the context or do we need to explicitly import certificateCommon.psm1 ?

iSazonov replied:
@kvprasoon You could insert
Import-Module (Join-Path -Path $PSScriptRoot 'certificateCommon.psm1') -Force

kvprasoon committed:
review comment fixes

kvprasoon committed:
random password generation

kvprasoon committed:
changes as per suggestion

kvprasoon commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -1,5 +1,8 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
+Import-Module (Join-Path -Path $PSScriptRoot '..\Microsoft.PowerShell.Security\certificateCommon.psm1') -Force
+Import-Module (Join-Path -Path $PSScriptRoot '..\..\..\Tools\Modules\HelpersCommon\HelpersCommon.psm1') -Force
```
addressed

kvprasoon committed:
Merge branch 'master' into srv-sdset

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Service.cs:
```diff
@@ -2688,6 +2738,15 @@ internal static extern
+
+        [DllImport(PinvokeDllNames.SetServiceObjectSecurityDllName, CharSet = CharSet.Unicode, SetLastError = true)]
+        internal static extern
+        bool SetServiceObjectSecurity(
```
@TravisEz13 Thanks for confirmation! (We already caught problems with Win7, Core and others)

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -1,5 +1,8 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
+Import-Module (Join-Path -Path $PSScriptRoot '..\Microsoft.PowerShell.Security\certificateCommon.psm1') -Force
```
Maybe move the needed function to HelpCommon.psm1?

PoshChan commented:
@TravisEz13, test results were not published for PowerShell-CI-linux at vstfs:///Build/Build/19649

PoshChan commented:
@TravisEz13, test results were not published for PowerShell-CI-macos at vstfs:///Build/Build/19648

PoshChan commented:
@TravisEz13, test results were not published for PowerShell-CI-static-analysis at vstfs:///Build/Build/19650

PoshChan commented:
@TravisEz13, your last commit had 2 failures in `PowerShell-CI-windows`
Set/New/Remove-Service cmdlet tests.Set-Service can change credentials of a service
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 183
183:             $service | Should -Not -BeNullOrEmpty
```
Set/New/Remove-Service cmdlet tests.Get-Service can get the 'UserName' of a service
```powershell
Expected a value, but got $null or empty.
at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 304
304:             $service | Should -Not -BeNullOrEmpty
```

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon commented:
Hope there is nothing pending from my side here.

Cc @SteveL-MSFT @TravisEz13  @PaulHigin

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -1,5 +1,8 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
+Import-Module (Join-Path -Path $PSScriptRoot '..\Microsoft.PowerShell.Security\certificateCommon.psm1')
+Import-Module HelpersCommon
```
Previously, we removed the explicit module loading and use autoloading.

kvprasoon replied:
Is `HelpersCommon` path there in the module path var in the CI agent machine ?

vexx32 replied:
I believe HelpersCommon is loaded when Start-PSPester is called.

kvprasoon replied:
Yep, `HelpersCommon` is loaded in `Start-PSPester`.

iSazonov replied:
Start-PSPester add path to test module directory in PSModulePath. So all test modules auto loaded.

kvprasoon replied:
yup, but `certificateCommon` is not. @iSazonov, Correct me if I'm wrong.

iSazonov replied:
It is here:
https://github.com/PowerShell/PowerShell/blob/83708a4242befaae0e26d7f33e7ebe2f0091db08/build.psm1#L171-L174

kvprasoon replied:
Modified.

kvprasoon committed:
removing explicit loading of HelpersCommon module

PoshChan commented:
@TravisEz13, could not find build at: , error: Invoke-RestMethod : {"$id":"1","innerException":null,"message":"The requested build 24524 could not be found.","typeName":"Microsoft.TeamFoundation.Build.WebApi.BuildNotFoundException, Microsoft.TeamFoundation.Build2.WebApi","typeKey":"BuildNotFoundException","errorCode":0,"eventId":3000}
At D:\home\site\wwwroot\Modules\AzDevOps\AzDevOps.psm1:7 char:9
+         Invoke-RestMethod -Uri $url -Authentication Basic -Credential ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : InvalidOperation: (Method: GET, Reques\u2026PowerShell/6.2.0
}:HttpRequestMessage) [Invoke-RestMethod], HttpResponseException
+ FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand

PoshChan commented:
@TravisEz13, this is the reminder you requested 1 hour ago

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon commented:
CI fails as no 7.x endpoints available, raised an issue.

https://github.com/PowerShell/PowerShell/issues/9999

PoshChan commented:
@TravisEz13, this is the reminder you requested 1 hour ago

kvprasoon commented:
@SteveL-MSFT CI is failing due to below error I think. I've closed #9999 

```
  Describing Set/New/Remove-Service cmdlet tests
The password does not meet the password policy requirements. Check the minimum password length, password complexity and password history requirements.

More help is available by typing NET HELPMSG 2245.

The password does not meet the password policy requirements. Check the minimum password length, password complexity and password history requirements.

More help is available by typing NET HELPMSG 2245.

The password does not meet the password policy requirements. Check the minimum password length, password complexity and password history requirements.

More help is available by typing NET HELPMSG 2245.

New-Service : Service ' (testsetcredential)' cannot be created due to the following error: The account name is invalid or does not exist, or the password is invalid for the account name specified.
At D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1:181 char:24
+             $service = New-Service @parameters
+                        ~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : PermissionDenied: (testsetcredential:String) [New-Service], ServiceCommandException
+ FullyQualifiedErrorId : CouldNotNewService,Microsoft.PowerShell.Commands.NewServiceCommand
 
The user name could not be found.

More help is available by typing NET HELPMSG 2221.

The user name could not be found.

More help is available by typing NET HELPMSG 2221.

++++++++++++++++++++++++++++
    [-] Set-Service can change credentials of a service 357ms
      Expected a value, but got $null or empty.
      182:             $service | Should -Not -BeNullOrEmpty
      at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 182
New-Service : Service ' (testgetservice)' cannot be created due to the following error: The account name is invalid or does not exist, or the password is invalid for the account name specified.
At D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1:302 char:24
+             $service = New-Service @startparameters
+                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : PermissionDenied: (testgetservice:String) [New-Service], ServiceCommandException
+ FullyQualifiedErrorId : CouldNotNewService,Microsoft.PowerShell.Commands.NewServiceCommand
 
++++++++
    [-] Get-Service can get the 'UserName' of a service 155ms
      Expected a value, but got $null or empty.
      303:             $service | Should -Not -BeNullOrEmpty
      at <ScriptBlock>, D:\a\1\s\test\powershell\Modules\Microsoft.PowerShell.Management\Set-Service.Tests.ps1: line 303
+++++++

    Context Set-Service test cases on the services with dependent relationship
The user name could not be found.

More help is available by typing NET HELPMSG 2221.
```

kvprasoon committed:
Merge branch 'master' into srv-sdset

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -8,14 +10,15 @@ Describe "Set/New/Remove-Service cmdlet tests" -Tags "Feature", "RequireAdminOnW
-            net user $userName $testPass /add > $null
-            $password = ConvertTo-SecureString $testPass -AsPlainText -Force
-            $creds = [pscredential]::new(".\$userName", $password)
+            $testPass = New-CertificatePassword
```
@kvprasoon It generate simple password based on New-RandomHexString. To fix tests you have to enhance New-CertificatePassword to generate complex password.

kvprasoon replied:
how about a new function

```
function New-ComplexPassword {    
    param(        
        [int]$Length            
    )
    $ASCII = 48..90 | Where-Object -FilterScript { $true} | ForEach-Object -Process { [char][byte]$_ }
    For ($Loop = 1; $Loop -le $Length; $Loop++) {
        
        $RandomPassword += ($ASCII | Get-Random)
        
    }    
$RandomPassword -as [string]
}
``

iSazonov replied:
You could grub a code from my PR (it creates passwords which works with the CI)
https://github.com/PowerShell/PowerShell/blob/34910f34734b1812c3106a202eff1e24215da107/test/powershell/Modules/Microsoft.PowerShell.AccountManagement/AccountManagement.User.Tests.ps1#L12-L13

kvprasoon replied:
Thanks @iSazonov - CI is now green

kvprasoon committed:
create complex password

kvprasoon committed:
Merge branch 'master' into srv-sdset

kvprasoon committed:
create complex pwd

kvprasoon commented on src/System.Management.Automation/utils/PInvokeDllNames.cs:
```diff
@@ -137,5 +137,6 @@ internal static class PinvokeDllNames
         internal const string DeleteServiceDllName = "api-ms-win-service-management-l1-1-0.dll";             /*124*/
         internal const string QueryServiceConfigDllName = "api-ms-win-service-management-l2-1-0.dll";        /*125*/
         internal const string QueryServiceConfig2DllName = "api-ms-win-service-management-l2-1-0.dll";       /*126*/
+        internal const string SetServiceObjectSecurityDllName = "api-ms-win-service-management-l2-1-0.dll";  /*127*/
```
Windows 10

kvprasoon commented on test/powershell/Modules/Microsoft.PowerShell.Management/Set-Service.Tests.ps1:
```diff
@@ -8,14 +10,16 @@ Describe "Set/New/Remove-Service cmdlet tests" -Tags "Feature", "RequireAdminOnW
+            $creds    = [pscredential]::new(".\$userName", $testPass)
+            $SecurityDescriptorSddl = 'D:(A;;CCLCSWRPWPDTLOCRRC;;;SY)(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;BA)(A;;CCLCSWLOCRRC;;;SU)'
+            $WrongSecurityDescriptorSddl = 'D:(A;;CCLCSWRPWPDTLOCRRC;;;SY)(A;;CCDCLCSWRPWPDTLOCRSDRCWDWO;;;BB)(A;;CCLCSWLOCRRC;;;SU)'
+            net user $userName $creds.GetNetworkCredential().Password /add > $null
```
It's already there in `afterall` block
